# Optimized-GeneCorr
Optimized gene correlation matrix generation from count data  
*Summer Research @ Bar-Ilan Bioinformatics Lab*

---

## Bioinformatics Lab Focus

**Overall Objective:**  
Analyze gene coexpression networks (GCNs) in elderly cells compared to younger cells, with the broader aim of applying these methodologies to cancer research. Similar to aging cells, cancer cells exhibit difficulties in regulating cell division and are increasingly prone to genetic mutations.  

**Methodology:**  
1. Develop a framework to build and analyze gene co-expression networks in scRNA-seq.  
2. Generate a cell-type-specific GCN atlas across young and aged tissues.  
3. Compare GCNs and their dynamics from young to aged tissues.  

**Bioinformatics significance:**  
Accurately building and analyzing coexpression networks is essential for identifying key genes, understanding their interactions, and drawing meaningful biological conclusions. 

**Network Analysis:**  
By constructing GCNs from gene expression data of both age groups, we can investigate structural and functional differences between the networks.  

**Features of Networks for Analysis:**  
- Graph density  
- Node expression  
- Edge co-expression  
- LCC size + density  
- Effective number of components  
- Degree assortativity  
- Percolation threshold  
- Node degree  
- Node harmonic


## Code Description 

**Coexpression.py** 
1. This does this
2. it also does this
3. it also does this
   
**Cython.py** 
1. this does this

**origional python code ** 
1. does this
2. does this does this
3. 


